### DISCUSSION TITLE
Developer Tips & Tricks

### DISCUSSION THREAD

**Discussion by profOnno** [helge17/tuxguitar#302] [General]
I started this discussion, because of my 'struggles' with using eclipse. At the moment my coding/figuring out configuration issues is at 20/80.

Things I learned so far:

After a sync of *your* GitHub branch (I do this on GitHub), then pull in this branch (I'm doing this from the command-line) there is a chance you need to update your maven project (Project->Update Maven project).

After the sync, i also need to update the run configuration. I delete the 'Java->Tuxguitar SWT' and use the import run configuration (from the development folder). This one bit me more than once. If you import the run configuration before you install the swt.jar in the `externals` folder. Things won't work.

**Comment by profOnno** [comment]
Tip for debugging plugins.
In the `run configurations`->`Dependencies` select `Classpath Enrtries` press `Add external Jars` browse to the `build-scripts/tuxguitar-linux-swt-x86_64/target/tuxguitar-9.99-SNAPSHOT-linux-swt-x86_64/share/plugins` select all of them and press `Open`

**Comment by profOnno** [comment]
Make sure to import the right maven project. The one in `tuxguitar/desktop/build_scripts/tuxguitar-linux-swt-x86_64`. In my eclipse version `2023-12`. After the import from git step I did close the dialog, instead of choosing one of the project imports. After this I did the import Maven from existing Maven project.

**Comment by guiv42** [comment]
Good idea to share :) This is where I am currently.

1. Build linux/SWT variant

I set up a build environment exactly following "INSTALL.md".
Then, from `/desktop/build-scripts/tuxguitar-linux-swt-x86_64` folder I build with command `mvn clean verify -P native-modules`. Or just `mvn verify` when I want to test minor modifications.

2. Eclipse

I did not strictly follow instructions from "contribute.md", since my environment was set up before this file was written.
Basically, open Eclipse, then "import existing Maven projects":
- desktop/build-scripts/tuxguitar-linux-swt-x86_64
- desktop/build-scripts/tuxguitar-linux-jfx-x86_64
- desktop/TuxGuitar-test/

Importing both SWT and JFX makes navigation possible in Eclipse in the 2 variants, which can be really useful when modifying code of user interface (I already modified the 2 implementations).
TuxGuitar-test is just because I cannot use JUnit/Maven correctly! I have implemented some tests, but as I wasn't quite sure of what I was doing I isolated them in this module. See #303: we should refactor this module.

3. Android

I imported manually the android project in AndroidStudio once, don't remember how :(
This definitely needs to be documented: I tried to re-import it after an upgrade of AndroidStudio but failed. Missing time for troubleshooting, so I restored my old config :(

4. Way of working

With this setup, I never succeeded to build TuxGuitar from Eclipse with the native modules. For an unknown reason, if I add "native-modules" in "Profiles" field of run configuration, I get the following error when it builds tuxguitar-alsa-linux:
`[exec] make: gcc: no such file or directory` (make and gcc are installed in my environment).
So I only use Eclipse to browse and edit the code.
Practically, I constantly keep a console open in `tuxguitar-linux-swt-x86_64` folder. I always build and run from this console.

Debugging is still an open issue, see #175 and #277.

I have regularly some errors raised in Eclipse, especially when I play around with git. Usually, just updating project solves the issue (Alt+F5, select all, OK).

Before creating a pull request, I try to build the app in different configurations. Not yet automated (still TODO), so I don't always rebuild all of them.
From a docker container (debian-based):
- Linux: (SWT, JFX) x (.tar.gz and .deb)
- Windows: SWT, JFX (standalone app only, not the installer)
From a virtual machine:
- FreeBSD: SWT, JFX
Manually from AndroidStudio:
- Android

When I find some time I will try to re-use the automated `misc/build_tuxguitar_from_source.sh` in my debian docker container.
